File: java/iroha_android/src/main/java/org/hyperledger/iroha/android/client/ToriiRequestBuilder.java
package org.hyperledger.iroha.android.client;

import java.net.URI;
import java.time.Duration;
import java.util.Map;
import java.util.Objects;
import org.hyperledger.iroha.android.client.transport.TransportRequest;
import org.hyperledger.iroha.android.norito.NoritoException;
import org.hyperledger.iroha.android.norito.SignedTransactionEncoder;
import org.hyperledger.iroha.android.tx.SignedTransaction;

/** Builds Torii HTTP requests for submitting signed transactions. */
final class ToriiRequestBuilder {
  private static final String SUBMIT_PATH = "/v1/pipeline/transactions";
  private static final String STATUS_PATH = "/v1/pipeline/transactions/status";

  private ToriiRequestBuilder() {}

  static TransportRequest buildSubmitRequest(
      final URI baseUri,
      final SignedTransaction transaction,
      final Duration timeout,
      final Map<String, String> extraHeaders) {
    Objects.requireNonNull(baseUri, "baseUri");
    Objects.requireNonNull(transaction, "transaction");
    final URI target = resolve(baseUri, SUBMIT_PATH);
    final byte[] norito;
    try {
      norito = SignedTransactionEncoder.encodeVersioned(transaction);
    } catch (final NoritoException ex) {
      throw new IllegalStateException("Failed to encode signed transaction", ex);
    }
    final TransportRequest.Builder builder =
        TransportRequest.builder()
            .setUri(target)
            .setMethod("POST")
            .addHeader("Content-Type", "application/x-norito")
            .addHeader("Accept", "application/x-norito, application/json")
            .setBody(norito);
    applyHeaders(builder, extraHeaders);
    applyTimeout(builder, timeout);
    return builder.build();
  }

  static TransportRequest buildStatusRequest(
      final URI baseUri,
      final String hashHex,
      final Duration timeout,
      final Map<String, String> extraHeaders) {
    Objects.requireNonNull(baseUri, "baseUri");
    final String normalizedHash =
        Objects.requireNonNull(hashHex, "hashHex").trim();
    if (normalizedHash.isEmpty()) {
      throw new IllegalArgumentException("hashHex must not be blank");
    }
    final URI target = resolve(baseUri, STATUS_PATH + "?hash=" + normalizedHash);
    final TransportRequest.Builder builder =
        TransportRequest.builder().setUri(target).setMethod("GET").addHeader("Accept", "application/json");
    applyHeaders(builder, extraHeaders);
    applyTimeout(builder, timeout);
    return builder.build();
  }

  private static URI resolve(final URI baseUri, final String path) {
    final String base = baseUri.toString();
    final String normalizedPath = path.startsWith("/") ? path.substring(1) : path;
    final String joined = base.endsWith("/")
        ? base + normalizedPath
        : base + "/" + normalizedPath;
    return URI.create(joined);
  }

  private static void applyHeaders(
      final TransportRequest.Builder builder, final Map<String, String> headers) {
    if (headers == null || headers.isEmpty()) {
      return;
    }
    for (Map.Entry<String, String> entry : headers.entrySet()) {
      if (entry.getKey() != null && entry.getValue() != null) {
        builder.addHeader(entry.getKey(), entry.getValue());
      }
    }
  }

  private static void applyTimeout(
      final TransportRequest.Builder builder, final Duration timeout) {
    if (timeout == null || timeout.isZero() || timeout.isNegative()) {
      return;
    }
    builder.setTimeout(timeout);
  }
}

File: kotlin/core-jvm/src/main/java/org/hyperledger/iroha/sdk/client/ToriiRequestBuilder.kt
package org.hyperledger.iroha.sdk.client

import java.net.URI
import java.time.Duration
import org.hyperledger.iroha.sdk.client.transport.TransportRequest
import org.hyperledger.iroha.sdk.tx.SignedTransaction
import org.hyperledger.iroha.sdk.tx.norito.NoritoException
import org.hyperledger.iroha.sdk.tx.norito.SignedTransactionEncoder

/** Builds Torii HTTP requests for submitting signed transactions. */
internal object ToriiRequestBuilder {
    private const val SUBMIT_PATH = "/transaction"
    private const val STATUS_PATH = "/v1/pipeline/transactions/status"

    @JvmStatic
    fun buildSubmitRequest(
        baseUri: URI,
        transaction: SignedTransaction,
        timeout: Duration?,
        extraHeaders: Map<String, String>?
    ): TransportRequest {
        val target = resolve(baseUri, SUBMIT_PATH)
        val norito: ByteArray
        try {
            norito = SignedTransactionEncoder.encodeVersioned(transaction)
        } catch (ex: NoritoException) {
            throw IllegalStateException("Failed to encode signed transaction", ex)
        }
        TransportSecurity.requireHttpRequestAllowed(
            "HttpClientTransport",
            baseUri,
            target,
            extraHeaders,
            norito,
        )
        val builder = TransportRequest.builder()
            .setUri(target)
            .setMethod("POST")
            .addHeader("Content-Type", "application/x-norito")
            .addHeader("Accept", "application/x-norito, application/json")
            .setBody(norito)
        applyHeaders(builder, extraHeaders)
        applyTimeout(builder, timeout)
        return builder.build()
    }

    @JvmStatic
    fun buildStatusRequest(
        baseUri: URI,
        hashHex: String,
        timeout: Duration?,
        extraHeaders: Map<String, String>?
    ): TransportRequest {
        val normalizedHash = hashHex.trim()
        require(normalizedHash.isNotEmpty()) { "hashHex must not be blank" }
        val target = resolve(baseUri, "$STATUS_PATH?hash=$normalizedHash")
        TransportSecurity.requireHttpRequestAllowed(
            "HttpClientTransport",
            baseUri,
            target,
            extraHeaders,
            null,
        )
        val builder = TransportRequest.builder()
            .setUri(target)
            .setMethod("GET")
            .addHeader("Accept", "application/json")
        applyHeaders(builder, extraHeaders)
        applyTimeout(builder, timeout)
        return builder.build()
    }

    private fun resolve(baseUri: URI, path: String): URI {
        val base = baseUri.toString()
        val normalizedPath = if (path.startsWith("/")) path.substring(1) else path
        val joined = if (base.endsWith("/")) base + normalizedPath else "$base/$normalizedPath"
        return URI.create(joined)
    }

    private fun applyHeaders(builder: TransportRequest.Builder, headers: Map<String, String>?) {
        if (headers.isNullOrEmpty()) return
        for ((key, value) in headers) {
            builder.addHeader(key, value)
        }
    }

    private fun applyTimeout(builder: TransportRequest.Builder, timeout: Duration?) {
        if (timeout == null || timeout.isZero || timeout.isNegative) return
        builder.setTimeout(timeout)
    }
}
